ee/src/main/java/org/jboss/as/ee/component/deployers/StartupCountdown.java

```diff
@@ -1,30 +1,45 @@
 package org.jboss.as.ee.component.deployers;
 
-import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.locks.AbstractQueuedSynchronizer;
 
 /**
  * Countdown tracker with capabilities similar to SE CountDownLatch, but allowing threads
- * to mark and unmark themselves as privileged. Privileged threads, when entering await method,
+ * to mark and unmark themselves as privileged and allowing counting up. Privileged threads, when entering await method,
  * will immediately proceed without checking latch's state. This reentrant behaviour allows to work around situations
  * where there is a possibility of a deadlock.
  * @author Fedor Gavrilov
  */
 public final class StartupCountdown {
-  private static final ThreadLocal<Boolean> isPrivileged = new ThreadLocal<Boolean>();
-
-  private final CountDownLatch latch;
-
-  public StartupCountdown(int count) {
-    this.latch = new CountDownLatch(count);
+  private final ThreadLocal<Boolean> isPrivileged = new ThreadLocal<Boolean>();
+  private final AbstractQueuedSynchronizer counter = new AbstractQueuedSynchronizer() {
```

Comment:
This looks overly complicated. Just use a volatile integer and double checked locking. It is easier to understand and should give better runtime performance because after startup the only overhead in await() is a volatile read.